### PR TITLE
Test Harness

### PR DESCRIPTION
**Pull request by ljtill**
**Matrix**
Terraform: 0.13.2, 0.13.5, 0.14.0, 0.14.6
Provider (Azure): 2.34.0, 2.46.1

**Steps**
Checkout / Install / Prepare / Init / Plan / Apply / Destroy

---

**Tasks**
- [x] Create pipeline manifests
- [x] Create Makefile controller
- [x] Create bash scripts
- [x] Create matrix coverage
- [x] Create default deployment
- [x] Create clean-up failback handler

### CODE REVIEW

File: tests/deployment/lib/archetype_definitions/archetype_definition_customer_sap.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (krowlandson) [inline tests/deployment/lib/archetype_definitions/archetype_definition_customer_sap.json:1]: We should probably remove this one and just specify the `default_empty` archetype definition ID where this is currently used.

File: tests/deployment/lib/archetype_definitions/archetype_definition_customer_landing_zone.tmpl.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (krowlandson) [inline tests/deployment/lib/archetype_definitions/archetype_definition_customer_landing_zone.tmpl.json:1]: We should probably remove this one and just specify the `default_empty` archetype definition ID where this is currently used.

File: tests/deployment/lib/archetype_definitions/archetype_definition_customer_corp.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (krowlandson) [inline tests/deployment/lib/archetype_definitions/archetype_definition_customer_corp.json:1]: We should probably remove this one and just specify the `default_empty` archetype definition ID where this is currently used.

**Review by krowlandson — CHANGES_REQUESTED** [review]
Looking great, thank you.

Let's run through the comments offline and get this merged.

File: tests/deployment/lib/archetype_definitions/archetype_definition_customer_corp.json (hunk 3 above)
Comment (krowlandson) [reply tests/deployment/lib/archetype_definitions/archetype_definition_customer_corp.json]: Updated in commit [Cleanup archetype definitions and variables](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pull/22/commits/cb736560540e4718a92e62ba256290f3382aa89f)

File: tests/deployment/lib/archetype_definitions/archetype_definition_customer_landing_zone.tmpl.json (hunk 2 above)
Comment (krowlandson) [reply tests/deployment/lib/archetype_definitions/archetype_definition_customer_landing_zone.tmpl.json]: Updated in commit [Cleanup archetype definitions and variables](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pull/22/commits/cb736560540e4718a92e62ba256290f3382aa89f)

File: tests/deployment/lib/archetype_definitions/archetype_definition_customer_sap.json (hunk 1 above)
Comment (krowlandson) [reply tests/deployment/lib/archetype_definitions/archetype_definition_customer_sap.json]: Updated in commit [Cleanup archetype definitions and variables](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pull/22/commits/cb736560540e4718a92e62ba256290f3382aa89f)

File: tests/scripts/tf-prepare.sh [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+echo "TF_ROOT_ID_1=$ROOT_ID_1" >> $GITHUB_ENV
+ROOT_ID_2=$RANDOM
+echo "==> Azure Root ID 2 - $ROOT_ID_2"
+echo "TF_ROOT_ID_2=$ROOT_ID_2" >> $GITHUB_ENV
+ROOT_ID_3=$RANDOM
+echo "==> Azure Root ID 3 - $ROOT_ID_3"
+echo "TF_ROOT_ID_3=$ROOT_ID_3" >> $GITHUB_ENV
+
+echo "==> Replacing provider version..."
+sed -i 's/version = ""/version = "'$TF_AZ_VERSION'"/g' main.tf
```
[the remark this answers is not in the document]
Comment (krowlandson) [reply tests/scripts/tf-prepare.sh:24]: Updated in commit [Replaced `version` with `required_providers` block](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pull/22/commits/71a3d80c2e5b0c6919b1d22af0201cf08e0dca22)

File: tests/deployment/lib/policy_assignments/policy_assignment_test_policy_set_definition.json [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+                "value": null
+            },
+            "CertificateThumbprints": {
+                "value": null
+            }
+        },
+        "policyDefinitionId": "/providers/Microsoft.Authorization/policySetDefinitions/a169a624-5599-4385-a696-c8d643089fab",
+        "scope": "${current_scope_resource_id}"
+    },
+    "sku": {
```
[the remark this answers is not in the document]
Comment (krowlandson) [reply tests/deployment/lib/policy_assignments/policy_assignment_test_policy_set_definition.json]: Found reference to this in the [ARM Template documentation for Policy Assignments](https://docs.microsoft.com/en-us/azure/templates/microsoft.authorization/policyassignments):

> The policy sku. This property is optional, obsolete, and will be ignored.

Have removed from the custom policy assignment templates in our test deployment, and will remove from all built-in policy assignment templates as part of the next policy refresh.

Update applied in commit [Remove obsolete `sku` block from templates](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pull/22/commits/01abb7575f6c2b6c595356a58dd6beb64d65bd96).

File: tests/scripts/tf-install.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+mkdir -p /home/runner/.terraform/bin/
+
+echo "==> Downloading archive..."
+wget 'https://releases.hashicorp.com/terraform/'$TF_VERSION'/terraform_'$TF_VERSION'_linux_amd64.zip' -P /tmp
+
+echo "==> Expanding archive..."
+unzip '/tmp/terraform_'$TF_VERSION'_linux_amd64.zip' -d /tmp
+
+echo "==> Moving binaries..."
+mv /tmp/terraform /home/runner/.terraform/bin/terraform
```
[the remark this answers is not in the document]
Comment (ljtill) [reply tests/scripts/tf-install.sh]: Correct - I've reviewed the test jobs and we have isolation between each version within the matrix.

File: tests/scripts/tf-init.sh [hunk 7]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/bash
+set -e
+
+#
+# Shell Script
+# - Terraform Initialize
+#
+
+echo "==> Switching directories..."
+cd $PWD/tests/deployment
```
[the remark this answers is not in the document]
Comment (ljtill) [reply tests/scripts/tf-init.sh]: Replaced the $PWD variable with a pipelines variables instead to better support this.

File: tests/scripts/tf-destroy.sh [hunk 8]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/bash
+set -e
+
+#
+# Shell Script
+# - Terraform Destroy
+#
+
+echo "==> Switching directories..."
+cd $PWD/tests/deployment
```
[the remark this answers is not in the document]
Comment (ljtill) [reply tests/scripts/tf-destroy.sh]: Replaced the $PWD variable with a pipelines variables instead to better support this.

**Review by krowlandson — APPROVED** [review]

**Pull request merged**